bindings/src/main/java/io/pravega/storage/filesystem/FileSystemBasedLock.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (c) 2019 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.storage.filesystem;
+
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Implements a mutually exclusive lock by using file system.
+ */
+public class FileSystemBasedLock implements AutoCloseable {
+    private static final long DEFAULT_SLEEP = Duration.of(10, ChronoUnit.MILLIS).toMillis();
+    private static final long DEFAULT_TIMEOUT = Duration.of(1, ChronoUnit.SECONDS).toMillis();
+    private static final long DEFAULT_LEASE_EXPIRATION = Duration.of(10, ChronoUnit.SECONDS).toMillis();
+
+    Path path;
+
+    /**
+     * Creates file system based lock.
+     * @param path Path to use for lock file.
+     * @throws Exception Throws exception.
+     */
+    public FileSystemBasedLock(Path path) throws Exception {
+        this(path, DEFAULT_LEASE_EXPIRATION, DEFAULT_SLEEP, DEFAULT_TIMEOUT);
+    }
+
+    /**
+     * Creates file system based lock.
+     * @param path Path to use for lock file.
+     * @param leaseDuration Lease expiration timeout.
+     * @param sleep Duartion to wait in mili.
+     * @param timeout Timeout in mili.
+     * @throws Exception Throws exception.
+     */
+    public FileSystemBasedLock(Path path, long leaseDuration, long sleep, long timeout) throws Exception {
+        long startTime = new Date().getTime();
+        this.path = path;
+        boolean done = false;
+        while (!done) {
+            try {
+                Files.createFile(path, PosixFilePermissions.asFileAttribute(FileSystemPermissions.READ_WRITE_PERMISSION));
+                return;
+            } catch (FileAlreadyExistsException ex) {
+                try {
+                    BasicFileAttributes attributes = Files.readAttributes(path, BasicFileAttributes.class);
+                    if (hasExpired(attributes.creationTime().toMillis(), leaseDuration)) {
```

Comment:
The lock relies on atomic nature of file create and delete. 
While acquiring lock we create a named file which is deleted when lock is released.
When lock is already held the file is present and any subsequent attempt to acquire lock will fail with FileAlreadyExistsException. 

However it is possible that process holding the lock might have died without releasing the lock hence leaving the zombie file behind. Here we are checking if lease has expired by checking the file creating time with current time.  The assumption is that lease expiry time is long enough that beyond which point it would be safe to assume original process has died.

Reply:
Say the expiration time is _x_, it doesn't matter what _x_ actually is. A process _p_ creates the file at time _t_, and say that it keeps running beyond _t + x_. If a second process _q_ comes trying to acquire the lock after _t + x_, it will delete the file and recreate it to acquire the lock, while _p_ will still think it holds the lock. Even if _p_ checks the lock, it will see the same file and will think that it still holds the lock. One way to get around it is to remember the creation time.

I believe this is making the assumption that an operation does not last for longer than the expiration period, which is a dangerous assumption, especially if we shorten the lease period to speed up recovery.

Also, leases typically have a time limit and are renewable, which is not the case for locks. If you want to do leases, then do leases, but do it properly with the ability of being renewed.